### PR TITLE
feat: add support for MADR v3

### PR DESCRIPTION
**Pull request by gaurav517**
## Hey, I just made a Pull Request!

Support parsing of [MADR v3 Format ](https://raw.githubusercontent.com/gaurav517/backstage/master/packages/catalog-model/examples/adrs/0003-madr-v3.md)

https://github.com/backstage/backstage/issues/16437

image

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: .changeset/rich-ladybugs-punch.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+'@backstage/plugin-adr-backend': minor
```
Comment (freben) [inline .changeset/rich-ladybugs-punch.md:2]: Nonbreaking changes in 0.x packages can be `patch`

**Review by freben — COMMENTED** [review]
Nice! Just a small tweak or two

**Review by kuangp — CHANGES_REQUESTED** [review]
My thoughts here is that maybe instead of overloading the default `mdrParser` to support 2 major versions, we ship separate parsers for MADRv2 and MADRv3. That way we can keep the parsing logic a little more organized and isolated. 

We can probably do this by updating [`MadrParserOptions`](https://github.com/backstage/backstage/blob/master/plugins/adr-backend/src/search/createMadrParser.ts#L42-L53) for `createMadrParser` to include an additional option to specify spec version and [apply the appropriate parser](https://github.com/backstage/backstage/blob/master/plugins/adr-backend/src/search/createMadrParser.ts#L67) accordingly. If we want to keep this a non-breaking change, we can default to using the current v2 parser implementation

File: ADOPTERS.md [hunk 2]
```diff
@@ -3,231 +3,232 @@
...
 _You can do this by using the [Adopter form](https://info.backstage.spotify.com/public-adopter-hand-raiser-form) or by editing this file after following the [CONTRIBUTING.md](./CONTRIBUTING.md)._
 
-| Organization                                                                | Contact                                                                                                                                                                                                                                                          | Description of Use                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     |
```
[the remark this answers is not in the document]
Comment (gaurav517) [reply ADOPTERS.md:6]: My editor auto formats! will fix it.

File: plugins/adr-backend/src/search/madrParser.ts [hunk 3]
```diff
@@ -31,22 +31,67 @@ export const madrParser = (content: string, dateFormat = MADR_DATE_FORMAT) => {
...
-
-  const adrDateTime = DateTime.fromFormat(
-    listTokens
-      ?.find(t => /^date:/i.test(t.text))
-      ?.text.replace(/^date:/i, '')
-      .trim() ?? '',
-    dateFormat,
-  );
+
+  let adrStatus: string | undefined;
```
[the remark this answers is not in the document]
Comment (gaurav517) [reply plugins/adr-backend/src/search/madrParser.ts:38]: Will fix it inline with @kuangp's suggestion.

**Comment by freben** [comment]
@kuangp Not to be selected hard-coded upfront though right? It'd be nice if we could register several parsers and they each get a go at the source data to detect whether they are able to understand it.

**Comment by gaurav517** [comment]
I pushed a changed based on @freben's feedback and did some related refactors.

About @kuangp's idea, there are few things to consider/discuss:
1. Looks like `madrParserOptions` is currently [not in use](https://github.com/backstage/backstage/blob/8006f8a602bb862c818715b8596d627fec7f296f/plugins/adr-backend/src/search/DefaultAdrCollatorFactory.ts#L117). We can use that and introduce an option for `version` but then it won't work well for organisations where some repos are using `v2` and some are using `v3` versions, as we would only be able to specify a specific version at backstage instance level: 
```
DefaultAdrCollatorFactory.fromConfig({
  adrVersions: v3 // this will break all v2 ADRs.
})
```
2. We can possibly overcome point-1 by having an additional annotation for adr version. e.g. `backstage.io/adr-version`, defaulting to `v2`. So that users can specify ADR versions at repo level. It would work assuming they won't be mixing versions in a repo. But thats a broader change in `DefaultAdrCollatorFactory`.

I am wondering whether we should just go ahead with current changes and come back to thinking about it later when we have another MADR version released? Supporting 2 versions by one parser is still not too bad atm. Please share your ideas. Thanks.

**Comment by kuangp** [comment]
> I am wondering whether we should just go ahead with current changes and come back to thinking about it later when we have another MADR version released? Supporting 2 versions by one parser is still not too bad atm. Please share your ideas. Thanks.

sure, I'm ok with that! @freben also makes a good point about potentially supporting a collection of parsers (not just MADR) to attempt to interpret the data but we can definitely revisit that later. Change looks good on my end, thanks for the contribution 🙂

**Review by kuangp — APPROVED** [review]

**Comment by gaurav517** [comment]
Great. Thanks. Then looks like its ready to ship. @freben :)

**Review by freben — APPROVED** [review]
Alright!

**Pull request merged**